### ISSUE TITLE
Could you implement cross-validation with callbacks when to stop?

### ISSUE THREAD

**Issue by Denisevi4** [Xtra-Computing/thundergbm#17]
and/or train tree by tree with callbacks whether to stop training?

that'd be nice.

**Comment by jiahuanluo** [comment]
Thanks for your advice. We are working on this.

**Comment by Denisevi4** [comment]
Just a suggestion. Xgboost creates K independent datasets for K-fold CV. You could save a lot of memory and make it fast if you instead set the id for each sample indicating which fold number is used in the testing.

For example if you pass a vector [0,1,2,3,4], then the first row would be used as a test in CV only in the first fold.

Or something like that. Do you understand what I mean?

**Comment by zhuyw05** [comment]
顶楼上，迫切希望补上callback和指定validate dataset的功能，这样就可以代替lightgbm了，速度真的快了很多很多，这样好的project不被完成太可惜了！  还有在使用的时候发现数据传输到GPU花的时间比训练长不少，要是能有builtin的grid search就更完美了

**Comment by zhuyw05** [comment]
没能力coding帮助楼主，楼主贴个支付宝吧我们赞助点咖啡钱

**Comment by zeyiwen** [comment]
@zhuyw05 Thank you for the feedback! We will work on that. Please stay tuned.

**Comment by GODqinbin** [comment]
Hi, we have added a function "cv" for cross-validation which supports the setting of the validation set. You can refer to [the document](https://github.com/Xtra-Computing/thundergbm/tree/master/python) for the parameters. We're still improving the python interface and will add callbacks such as early stopping in the future. Thanks.
